File: workspace/src/dose_sammler/scripts/encoder_odometry.cpp
#include <ros/ros.h>
#include <nav_msgs/Odometry.h>
#include <sensor_msgs/JointState.h>
#include <geometry_msgs/TransformStamped.h>
#include <tf2_ros/transform_broadcaster.h>
#include <tf2/LinearMath/Quaternion.h>
#include <pigpio.h>
#include <cmath>
#include <map>
#include <string>

// Encoder pins
const std::map<std::string, std::pair<int, int>> ENCODER_PINS = {
    {"FL", {19, 13}},
    {"FR", {11, 0}},
    {"RL", {22, 27}},
    {"RR", {24, 23}}
};

// Tick counter
std::map<std::string, int> ticks = {{"FL", 0}, {"FR", 0}, {"RL", 0}, {"RR", 0}};
std::map<std::string, int> last_states;

// Callback function for encoder interrupts
void encoderCallback(int gpio, int level, uint32_t tick, const std::string& wheel_name, int pin_a, int pin_b) {
    int a = gpioRead(pin_a);
    int b = gpioRead(pin_b);
    int current_state = (a << 1) | b;

    int last_state = last_states[wheel_name];
    int transition = (last_state << 2) | current_state;

    // Valid transitions for forward and backward movement
    if (transition == 0b0001 || transition == 0b0111 || transition == 0b1110 || transition == 0b1000) {
        ticks[wheel_name]++;
    } else if (transition == 0b0010 || transition == 0b0100 || transition == 0b1101 || transition == 0b1011) {
        ticks[wheel_name]--;
    }

    last_states[wheel_name] = current_state;
}

// Setup encoders
void setupEncoders() {
    if (gpioInitialise() < 0) {
        ROS_ERROR("Failed to initialize pigpio library");
        ros::shutdown();
        return;
    }

    for (const auto& [wheel_name, pins] : ENCODER_PINS) {
        int pin_a = pins.first;
        int pin_b = pins.second;

        gpioSetMode(pin_a, PI_INPUT);
        gpioSetMode(pin_b, PI_INPUT);
        gpioSetPullUpDown(pin_a, PI_PUD_UP);
        gpioSetPullUpDown(pin_b, PI_PUD_UP);

        last_states[wheel_name] = (gpioRead(pin_a) << 1) | gpioRead(pin_b);

        gpioSetAlertFuncEx(pin_a, [](int gpio, int level, uint32_t tick, void* userdata) {
            auto* data = static_cast<std::pair<std::string, std::pair<int, int>>*>(userdata);
            encoderCallback(gpio, level, tick, data->first, data->second.first, data->second.second);
        }, new std::pair<std::string, std::pair<int, int>>(wheel_name, pins));

        gpioSetAlertFuncEx(pin_b, [](int gpio, int level, uint32_t tick, void* userdata) {
            auto* data = static_cast<std::pair<std::string, std::pair<int, int>>*>(userdata);
            encoderCallback(gpio, level, tick, data->first, data->second.first, data->second.second);
        }, new std::pair<std::string, std::pair<int, int>>(wheel_name, pins));
    }
}

// Calculate odometry
std::tuple<double, double, double, std::map<std::string, double>> calculateOdometry(const std::map<std::string, int>& ticks_delta, double dt) {
    const double TICKS_PER_REV = 2797;
    const double WHEEL_RADIUS = 0.044;
    const double L = 0.244;
    const double W = 0.132;

    std::map<std::string, double> wheel_speeds;
    double wheel_circumference = 2 * M_PI * WHEEL_RADIUS;

    for (const auto& [wheel_name, delta] : ticks_delta) {
        double revs = delta / TICKS_PER_REV;
        wheel_speeds[wheel_name] = (wheel_circumference * revs) / dt;
    }

    double vx = (wheel_speeds["FL"] + wheel_speeds["FR"] + wheel_speeds["RL"] + wheel_speeds["RR"]) / 4;
    double vy = (-wheel_speeds["FL"] + wheel_speeds["FR"] + wheel_speeds["RL"] - wheel_speeds["RR"]) / 4;
    double wz = (-wheel_speeds["FL"] - wheel_speeds["FR"] + wheel_speeds["RL"] + wheel_speeds["RR"]) / (4 * (L + W));

    return {vx, vy, wz, wheel_speeds};
}

// Main function
int main(int argc, char** argv) {
    ros::init(argc, argv, "encoder_odometry");
    ros::NodeHandle nh;

    ros::Publisher odom_pub = nh.advertise<nav_msgs::Odometry>("/odom", 50);
    ros::Publisher joint_pub = nh.advertise<sensor_msgs::JointState>("/wheel_states", 10);
    tf2_ros::TransformBroadcaster tf_broadcaster;

    setupEncoders();

    double x = 0.0, y = 0.0, th = 0.0;
    std::map<std::string, int> last_ticks = ticks;
    ros::Time last_time = ros::Time::now();
    ros::Rate rate(100); // 100 Hz

    try {
        while (ros::ok()) {
            ros::Time now = ros::Time::now();
            double dt = (now - last_time).toSec();
            last_time = now;

            if (dt == 0) {
                rate.sleep();
                continue;
            }

            std::map<std::string, int> ticks_delta;
            for (const auto& [wheel_name, tick] : ticks) {
                ticks_delta[wheel_name] = tick - last_ticks[wheel_name];
            }
            last_ticks = ticks;

            auto [vx, vy, vth, wheel_speeds] = calculateOdometry(ticks_delta, dt);

            double delta_x = (vx * cos(th) - vy * sin(th)) * dt;
            double delta_y = (vx * sin(th) + vy * cos(th)) * dt;
            double delta_th = vth * dt;

            x += delta_x;
            y += delta_y;
            th += delta_th;

            geometry_msgs::TransformStamped t;
            t.header.stamp = now;
            t.header.frame_id = "odom";
            t.child_frame_id = "base_link";
            t.transform.translation.x = -x;
            t.transform.translation.y = y;
            t.transform.translation.z = 0.0;

            // Fix quaternion calculation
            tf2::Quaternion q;
            q.setRPY(0, 0, th);
            t.transform.rotation.x = q.x();
            t.transform.rotation.y = q.y();
            t.transform.rotation.z = q.z();
            t.transform.rotation.w = q.w();
            tf_broadcaster.sendTransform(t);

            nav_msgs::Odometry odom;
            odom.header.stamp = now;
            odom.header.frame_id = "odom";
            odom.pose.pose.position.x = x;
            odom.pose.pose.position.y = y;
            odom.pose.pose.position.z = 0.0;
            odom.pose.pose.orientation = t.transform.rotation;
            odom.child_frame_id = "base_link";
            odom.twist.twist.linear.x = vx;
            odom.twist.twist.linear.y = vy;
            odom.twist.twist.angular.z = vth;
            odom_pub.publish(odom);

            sensor_msgs::JointState joint_state;
            joint_state.header.stamp = now;
            joint_state.name = {"FL", "FR", "RL", "RR"};
            joint_state.position = {0.0, 0.0, 0.0, 0.0};
            joint_state.velocity = {wheel_speeds["FL"], wheel_speeds["FR"], wheel_speeds["RL"], wheel_speeds["RR"]};
            joint_state.effort = {0.0, 0.0, 0.0, 0.0};
            joint_pub.publish(joint_state);

            rate.sleep();
        }
    } catch (const std::exception& e) {
        ROS_ERROR("Exception: %s", e.what());
    }

    gpioTerminate();
    return 0;
}

File: workspace/src/dose_sammler/scripts/encoder_odometry.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
Created on Wed Apr 30 13:46:31 2025

@author: Project Curia
"""

import rospy
import RPi.GPIO as GPIO
import math
from nav_msgs.msg import Odometry
from geometry_msgs.msg import Quaternion, Twist, TransformStamped
from tf.transformations import quaternion_from_euler
import tf2_ros

from sensor_msgs.msg import JointState
from std_msgs.msg import Int32MultiArray

# Define the pins for the encoder:
ENCODER_PINS = {
    "FL_A": 22, "FL_B": 27,
    "FR_A": 24, "FR_B": 23,
    "RL_A": 19, "RL_B": 13,
    "RR_A": 11, "RR_B": 0
}

# Tick counter:
import threading
ticks = {"FL": 0, "FR": 0, "RL": 0, "RR": 0}
ticks_lock = threading.Lock()

# Store direction for each wheel (default to 1)
encoder_dir = {"FL": 1, "FR": 1, "RL": 1, "RR": 1}
encoder_dir_lock = threading.Lock()

# Callback for direction topic
def dir_callback(msg):
    # msg should be a dictionary-like message, e.g. {'FL': 1, 'FR': 1, 'RL': -1, 'RR': -1}
    # For example, use std_msgs/Int32MultiArray or a custom message
    with encoder_dir_lock:
        encoder_dir['FL'] = msg.data[0]
        encoder_dir['FR'] = msg.data[1]
        encoder_dir['RL'] = msg.data[2]
        encoder_dir['RR'] = msg.data[3]

# Unified encoder callback
def encoder_callback(channel, wheel_name):
    # Only increment/decrement on rising edge of encoder A
    with encoder_dir_lock:
        dir = encoder_dir[wheel_name]
    with ticks_lock:
        ticks[wheel_name] += dir

# -------------- UPDATED SETUP FUNCTION --------------


def setup_encoders():
    GPIO.setmode(GPIO.BCM)
    # Define wheel info: (A pin, direction)
    wheel_info = {
        "FL": (ENCODER_PINS["FL_A"],  1),
        "FR": (ENCODER_PINS["FR_A"],  1),
        "RL": (ENCODER_PINS["RL_A"], -1),
        "RR": (ENCODER_PINS["RR_A"], -1),
    }
    for wheel, (a_pin, dir) in wheel_info.items():
        GPIO.setup(a_pin, GPIO.IN, pull_up_down=GPIO.PUD_UP)
        # Use lambda to pass wheel_name and dir to the callback
        GPIO.add_event_detect(
            a_pin, GPIO.RISING,
            callback=lambda channel, w=wheel, d=dir: encoder_callback(channel, w, d)
        )


# Define the function to calculate the odometry:
def calculate_odometry(ticks_delta, dt):
    TICKS_PER_REV = 700 #2797
    WHEEL_RADIUS = 0.044
    L = 0.244
    W = 0.132

    w = {}

    wheel_circumference = 2 * math.pi * WHEEL_RADIUS
    for name in ticks_delta:
        revs = ticks_delta[name] / TICKS_PER_REV
        w[name] = (wheel_circumference * revs) / dt

    # Inverse kinematics:
    vx = (w["FL"] + w["FR"] + w["RL"] + w["RR"]) / 4
    vy = (-w["FL"] + w["FR"] + w["RL"] - w["RR"]) / 4
    wz = (-w["FL"] - w["FR"] + w["RL"] + w["RR"]) / (4 * (L + W))

    return vx, vy, wz, w


# Define the main function:
def main():
    # Init the node:
    rospy.init_node("encoder_odometry")
    # Unterhalb von odom_pub diese Zeile einfuegen:
    joint_pub = rospy.Publisher("/wheel_speeds", JointState, queue_size=10)
    odom_pub = rospy.Publisher("/odom", Odometry, queue_size=50)
    tf_broadcaster = tf2_ros.TransformBroadcaster()

    # Subscribe to direction topic (example: Int32MultiArray)
    rospy.Subscriber("/wheel_direction", Int8MultiArray, dir_callback)

    # Setup the encoders:
    setup_encoders()

    # Define the rate and further needed parameters:
    rate = rospy.Rate(100)  # 100 Hz
    x = y = th = 0.0
    last_ticks = ticks.copy()
    last_time = rospy.Time.now()

    while not rospy.is_shutdown():
        # Calculate the passed time:
        now = rospy.Time.now()
        dt = (now - last_time).to_sec()
    
        if dt == 0:
            rate.sleep()
            continue

        # Calculate the amount of ticks:
        ticks_delta = {k: ticks[k] - last_ticks[k] for k in ticks}
        last_ticks = ticks.copy()
        
        # Calculate the odometry:
        vx, vy, vth, w = calculate_odometry(ticks_delta, dt)

        # Calculate dx, dy and drot:
        delta_x = (vx * math.cos(th) - vy * math.sin(th)) * dt
        delta_y = (vx * math.sin(th) + vy * math.cos(th)) * dt
        delta_th = vth * dt

        x += delta_x
        y += delta_y
        th += delta_th

        # Calculate the quaternion from the parameters:
        odom_quat = quaternion_from_euler(0, 0, th)

        # Define the transformation:
        t = TransformStamped()
        t.header.stamp = now
        t.header.frame_id = "odom"
        t.child_frame_id = "base_link"
        t.transform.translation.x = -x
        t.transform.translation.y = y
        t.transform.translation.z = 0.0
        t.transform.rotation.x = odom_quat[0]
        t.transform.rotation.y = odom_quat[1]
        t.transform.rotation.z = odom_quat[2]
        t.transform.rotation.w = odom_quat[3]
        tf_broadcaster.sendTransform(t)

        # Get the odometry:
        odom = Odometry()
        odom.header.stamp = now
        odom.header.frame_id = "odom"
        odom.pose.pose.position.x = x
        odom.pose.pose.position.y = y
        odom.pose.pose.position.z = 0.0
        odom.pose.pose.orientation.x = odom_quat[0]
        odom.pose.pose.orientation.y = odom_quat[1]
        odom.pose.pose.orientation.z = odom_quat[2]
        odom.pose.pose.orientation.w = odom_quat[3]

        odom.child_frame_id = "base_link"
        odom.twist.twist.linear.x = vx
        odom.twist.twist.linear.y = vy
        odom.twist.twist.angular.z = vth

        # JointState Nachricht mit Ticks & Radgeschwindigkeit
        joint_state = JointState()
        joint_state.header.stamp = now
        joint_state.name = ['FL', 'FR', 'RL', 'RR']
        joint_state.position = [ticks['FL'], ticks['FR'], ticks['RL'], ticks['RR']]
        joint_state.velocity = [w['FL'], w['FR'], w['RL'], w['RR']]
        joint_state.effort = [0.0, 0.0, 0.0, 0.0]

        joint_pub.publish(joint_state)

        # Publish the odom:
        odom_pub.publish(odom)
        last_time = now
        rate.sleep()

# Call the main function if it's executed as the main file:
if __name__ == "__main__":
    try:
        main()
    except rospy.ROSInterruptException:
        GPIO.cleanup()
